transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapPacket.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.channel.epoll;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+
+/**
+ * Represents a packet being received or sent via a {@link TunTapChannel}.
+ */
+public final class TunTapPacket implements ByteBufHolder {
+
+    private int protocol;
+    private ByteBuf packetData;
+
+    public TunTapPacket() {
+        // Nothing to do.
+    }
+
+    public TunTapPacket(int protocol, ByteBuf packetData) {
+        if (packetData == null) {
+            throw new NullPointerException("packetData");
+        }
+        this.protocol = protocol;
+        this.packetData = packetData;
+    }
+
+    public TunTapPacket(ByteBuf packetData) {
+        if (packetData == null) {
+            throw new NullPointerException("packetData");
+        }
+        protocol = inferProtocolFromIPPacket(packetData);
+        this.packetData = packetData;
+    }
+
+    public int protocol() {
+        return protocol;
+    }
+
+    public void setProtocol(int protocol) {
```

Comment:
consider removing mutators unless they are necessary, and then make the member variables `final`.